generate_engagement_analysis_files.py

```diff
@@ -0,0 +1,152 @@
+            demog = data[uid]
+
+            if uid not in engagement_map:
+                engagement_map[uid] = {
+                    "demog": demog,
+                    "shows": []
+                }
+                assert demogs == engagement_map[uid]['demog'] , f"{demogs} not equal to {engagement_map[uid]['demog']}" \
+                    f" for {uid}"
+
+            engagement_map[uid]["shows"].append(rqa_raw_field)
+            total_participants_with_optins.add(uid)
+            total_activations_with_optins.append(uid)
+
+    # Export the engagement counts to their respective csv file.
+    log.info(f'Writing total participants with optins csv ...')
+    with open(f"{engagement_csv_output_dir}/total_participants_with_optins.csv", "w") as f:
+        writer = csv.writer(f)
+        writer.writerow([len(total_participants_with_optins)])
+
+    log.info(f'Writing total activations with opt-ins csv ...')
+    with open(f"{engagement_csv_output_dir}/total_activations_with_optins.csv", "w") as f:
+        writer = csv.writer(f)
+        writer.writerow([len(total_activations_with_optins)])
+
+    log.info(f'Writing activations_per_show_with_optins ...')
+    with open(f"{engagement_csv_output_dir}/activations_per_show_with_optins.csv", "w") as f:
+        writer = csv.writer(f)
+
+        for row in activations_per_show_with_optins.items():
+            writer.writerow(row)
+
+    # For each uid generate a sustained engagement map that contains : no of shows participated , their manually
+    # coded demographics and a matrix representation of the shows they have participated in
+    # (1 for show they participated and 0 otherwise.)
+    sustained_engagement_map = {}
+    for uid in engagement_map.keys():
+        basic_str = f" {len(engagement_map[uid]['shows'])}, {engagement_map[uid]['demog']['gender']}," \
+            f" {engagement_map[uid]['demog']['age']}, {engagement_map[uid]['demog']['recently_displaced']}, "
+
+        show_mapping = ""
+        for show_name in rqa_raw_fields:
+            if show_name in engagement_map[uid]["shows"]:
+                show_mapping = show_mapping + "1, "
+            else:
+                show_mapping = show_mapping + "0, "
+
+        compound_str = basic_str + show_mapping
+        sustained_engagement_map[uid] = [compound_str]
+
+    log.info(f'Computing repeat engagement / total_participants_with_optins ...' )
+    # Compute the number of consented individuals who participated exactly 1 to <number of RQAs> times.
+    # A consented individual is considered to have participated if they sent a message, regardless of the
+    # relevance of any of their messages.
+    repeat_engagement = OrderedDict()
+    for i in range(1, len(rqa_raw_fields) + 1):
+        repeat_engagement[i] = {
+            "Repeat No": i,
+            "No. of participants with opt-ins": 0,
+            "% of participants with opt-ins": None
+        }
+        for k, v in sustained_engagement_map.items():
```

Comment:
Can you give these more useful names - I have to go back up to remember what k and v are

Reply:
renamed k -> uid and v -> compound_str